File: backend/app/models/inventory.py
from datetime import datetime
from sqlalchemy import (
    Integer, String, Boolean, Text, DateTime, Numeric,
    ForeignKey, func
)
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class InventoryProduct(Base):
    __tablename__ = "inventory_products"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    sku: Mapped[str | None] = mapped_column(String(60), unique=True, nullable=True)
    category: Mapped[str] = mapped_column(String(60), nullable=False)  # extensions/tools/retail/color/care
    supplier_name: Mapped[str | None] = mapped_column(String(120), nullable=True)
    supplier_contact: Mapped[str | None] = mapped_column(String(255), nullable=True)
    unit_cost: Mapped[float | None] = mapped_column(Numeric(8, 2), nullable=True)
    retail_price: Mapped[float | None] = mapped_column(Numeric(8, 2), nullable=True)
    current_stock: Mapped[float] = mapped_column(Numeric(10, 2), default=0)
    stock_unit: Mapped[str] = mapped_column(String(20), default="units")  # units/grams/packs
    reorder_threshold: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    reorder_quantity: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    last_ordered_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    last_restocked_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=func.now(), onupdate=func.now()
    )

    # Relationships
    transactions: Mapped[list["InventoryTransaction"]] = relationship(
        "InventoryTransaction", back_populates="product", cascade="all, delete-orphan"
    )


class InventoryTransaction(Base):
    __tablename__ = "inventory_transactions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("inventory_products.id"), nullable=False
    )
    transaction_type: Mapped[str] = mapped_column(
        String(20), nullable=False
    )  # received/used/adjusted/sold/wasted
    quantity_change: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    quantity_after: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    appointment_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("appointments.id"), nullable=True
    )
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=func.now())

    product: Mapped["InventoryProduct"] = relationship(
        "InventoryProduct", back_populates="transactions"
    )
    appointment: Mapped["Appointment | None"] = relationship(  # type: ignore[name-defined]  # noqa
        "Appointment", back_populates="inventory_transactions"
    )


class PurchaseOrder(Base):
    __tablename__ = "purchase_orders"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    status: Mapped[str] = mapped_column(
        String(20), default="draft"
    )  # draft/sent/received/cancelled
    supplier_name: Mapped[str | None] = mapped_column(String(120), nullable=True)
    ai_generated: Mapped[bool] = mapped_column(Boolean, default=False)
    items_json: Mapped[str] = mapped_column(Text, nullable=False)  # JSON array
    total_cost: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    ordered_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    received_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=func.now())

File: frontend/src/api/inventory.ts
import api from "./client";

export interface InventoryProduct {
  id: number;
  sku?: string;
  name: string;
  category?: string;
  supplier_name?: string;
  unit_cost?: number;
  retail_price?: number;
  current_stock: number;
  stock_unit: string;
  reorder_threshold: number;
  reorder_quantity: number;
  is_active: boolean;
  last_ordered_at?: string;
}

export interface ProductCreate {
  sku?: string;
  name: string;
  category?: string;
  supplier_name?: string;
  supplier_sku?: string;
  supplier_contact?: string;
  unit_cost?: number;
  retail_price?: number;
  current_stock?: number;
  stock_unit?: string;
  reorder_threshold?: number;
  reorder_quantity?: number;
}

export interface PurchaseOrder {
  id: number;
  status: "draft" | "sent" | "received" | "cancelled";
  supplier_name: string;
  ai_generated: boolean;
  items_json: unknown[];
  total_cost?: number;
  ordered_at?: string;
  received_at?: string;
  created_at: string;
}

export const inventoryApi = {
  listProducts: (params?: { category?: string; low_stock?: boolean }) =>
    api
      .get<InventoryProduct[]>("/inventory/products", { params })
      .then((r) => r.data),

  getProduct: (id: number) =>
    api.get<InventoryProduct>(`/inventory/products/${id}`).then((r) => r.data),

  createProduct: (data: ProductCreate) =>
    api.post<InventoryProduct>("/inventory/products", data).then((r) => r.data),

  updateProduct: (id: number, data: Partial<ProductCreate>) =>
    api
      .put<InventoryProduct>(`/inventory/products/${id}`, data)
      .then((r) => r.data),

  adjustStock: (
    id: number,
    data: { transaction_type: string; quantity_change: number; notes?: string }
  ) =>
    api
      .post(`/inventory/products/${id}/adjust`, data)
      .then((r) => r.data),

  getAlerts: () => api.get("/inventory/alerts").then((r) => r.data),

  getReorderAdvice: () =>
    api.post("/inventory/reorder-advice").then((r) => r.data),

  listOrders: () =>
    api.get<PurchaseOrder[]>("/inventory/purchase-orders").then((r) => r.data),

  createOrder: (data: {
    supplier_name: string;
    items_json: unknown[];
    ai_generated?: boolean;
  }) =>
    api
      .post<PurchaseOrder>("/inventory/purchase-orders", data)
      .then((r) => r.data),

  updateOrder: (id: number, data: Partial<PurchaseOrder>) =>
    api
      .put<PurchaseOrder>(`/inventory/purchase-orders/${id}`, data)
      .then((r) => r.data),
};
